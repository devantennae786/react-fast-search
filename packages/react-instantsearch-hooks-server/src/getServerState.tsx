import { isIndexWidget } from 'instantsearch.js/es/widgets/index/index';
import React from 'react';
import ReactDOM from 'react-dom/server';
import { InstantSearchServerContext } from 'react-instantsearch-hooks';

import type { InitialResults, InstantSearch } from 'instantsearch.js';
import type { IndexWidget } from 'instantsearch.js/es/widgets/index/index';
import type { ReactNode } from 'react';
import type {
  InstantSearchServerContextApi,
  InstantSearchServerState,
} from 'react-instantsearch-hooks';

/**
 * Returns the InstantSearch server state from a component.
 */
export async function getServerState(
  children: ReactNode
): Promise<InstantSearchServerState> {
  const searchRef: { current: InstantSearch | undefined } = {
    current: undefined,
  };

  const notifyServer: InstantSearchServerContextApi['notifyServer'] = ({
    search,
  }) => {
    searchRef.current = search;
  };

  ReactDOM.renderToString(
    <InstantSearchServerContext.Provider value={{ notifyServer }}>
      {children}
    </InstantSearchServerContext.Provider>
  );

  // We wait for the component to mount so that `notifyServer()` is called.
  await new Promise((resolve) => setTimeout(resolve, 0));

  // If `notifyServer()` is not called by then, it means that <InstantSearch>
  // wasn't within the `children`.
  // We decide to go with a strict behavior in that case; throwing. If users have
  // some routes that don't mount the <InstantSearch> component, they would need
  // to try/catch the `getServerState()` call.
  // If this behavior turns out to be too strict for many users, we can decide
  // to warn instead of throwing.
  if (!searchRef.current) {
    throw new Error(
      "Unable to retrieve InstantSearch's server state in `getServerState()`. Did you mount the <InstantSearch> component?"
    );
  }

  await waitForResults(searchRef.current);

  const initialResults = getInitialResults(searchRef.current.mainIndex);

  return {
    initialResults,
  };
}

/**
 * Waits for the results from the search instance to coordinate the next steps
 * in `getServerState()`.
 */
function waitForResults(search: InstantSearch) {
  const helper = search.mainHelper!;

  helper.searchOnlyWithDerivedHelpers();

  return new Promise<void>((resolve, reject) => {
    // All derived helpers resolve in the same tick so we're safe only relying
    // on the first one.
    helper.derivedHelpers[0].on('result', () => {
      resolve();
    });

    // However, we listen to errors that can happen on any derived helper because
    // any error is critical.
    helper.derivedHelpers.forEach((derivedHelper) =>
      derivedHelper.on('error', (error) => {
        reject(error);
      })
    );
  });
}

/**
 * Walks the InstantSearch root index to construct the initial results.
 */
function getInitialResults(rootIndex: IndexWidget): InitialResults {
  function walkIndex(
    indexWidget: IndexWidget,
    callback: (widget: IndexWidget) => void
  ) {
    callback(indexWidget);

    return indexWidget.getWidgets().forEach((widget) => {
      if (!isIndexWidget(widget)) {
        return;
      }

      callback(widget);
      walkIndex(widget, callback);
    });
  }

  const initialResults: InitialResults = {};

  walkIndex(rootIndex, (widget) => {
    const searchResults = widget.getResults()!;
    initialResults[widget.getIndexId()] = {
      // We convert the Helper state to a plain object to pass parsable data
      // structures from server to client.
      state: { ...searchResults._state },
      results: searchResults._rawResults,
    };
  });

  return initialResults;
}
