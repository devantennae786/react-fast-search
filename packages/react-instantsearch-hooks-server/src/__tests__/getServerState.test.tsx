/**
 * @jest-environment node
 */

import React, { version as ReactVersion } from 'react';
import {
  InstantSearch,
  InstantSearchSSRProvider,
  Index,
  DynamicWidgets,
  useHits,
  useRefinementList,
  useSearchBox,
  version,
} from 'react-instantsearch-hooks';

import { createSearchClient } from '../../../../test/mock';
import { getServerState } from '../getServerState';

import type {
  InstantSearchServerState,
  InstantSearchProps,
  UseRefinementListProps,
} from 'react-instantsearch-hooks';

type CreateTestEnvironmentProps = {
  searchClient: InstantSearchProps['searchClient'];
  initialUiState?: InstantSearchProps['initialUiState'];
};

function createTestEnvironment({
  searchClient,
  initialUiState = {
    instant_search: {
      query: 'iphone',
      refinementList: {
        brand: ['Apple'],
      },
    },
    instant_search_price_asc: {
      query: 'iphone',
      refinementList: {
        brand: ['Apple'],
      },
    },
  },
}: CreateTestEnvironmentProps) {
  function Search({ children }: { children?: React.ReactNode }) {
    return (
      <InstantSearch
        searchClient={searchClient}
        indexName="instant_search"
        initialUiState={initialUiState}
      >
        {children}
        <RefinementList attribute="brand" />
        <SearchBox />
        <Hits />

        <Index indexName="instant_search_price_asc">
          <Hits />

          <Index indexName="instant_search_rating_desc">
            <Hits />
          </Index>
        </Index>

        <Index indexName="instant_search_price_desc">
          <Hits />
        </Index>
      </InstantSearch>
    );
  }

  function App({
    serverState,
    children,
  }: {
    serverState?: InstantSearchServerState;
    children?: React.ReactNode;
  }) {
    return (
      <InstantSearchSSRProvider {...serverState}>
        <Search>{children}</Search>
      </InstantSearchSSRProvider>
    );
  }

  return {
    App,
  };
}

describe('getServerState', () => {
  test('throws without <InstantSearch> component', async () => {
    function App() {
      return null;
    }

    await expect(
      getServerState(<App />)
    ).rejects.toThrowErrorMatchingInlineSnapshot(
      `"Unable to retrieve InstantSearch's server state in \`getServerState()\`. Did you mount the <InstantSearch> component?"`
    );
  });

  test('throws when the search client errors', async () => {
    const searchClient = createSearchClient({
      search: () => {
        throw new Error('Search client error');
      },
    });
    const { App } = createTestEnvironment({ searchClient });

    await expect(getServerState(<App />)).rejects.toThrow(
      /Search client error/
    );
  });

  test('adds the server user agents', async () => {
    const searchClient = createSearchClient();
    const { App } = createTestEnvironment({ searchClient });

    await getServerState(<App />);

    expect(searchClient.addAlgoliaAgent).toHaveBeenCalledWith(
      `react (${ReactVersion})`
    );
    expect(searchClient.addAlgoliaAgent).toHaveBeenCalledWith(
      `react-instantsearch (${version})`
    );
    expect(searchClient.addAlgoliaAgent).toHaveBeenCalledWith(
      `react-instantsearch-hooks (${version})`
    );
    expect(searchClient.addAlgoliaAgent).toHaveBeenCalledWith(
      `react-instantsearch-server (${version})`
    );
  });

  test('calls search with widgets search parameters', async () => {
    const searchClient = createSearchClient();
    const { App } = createTestEnvironment({ searchClient });

    await getServerState(<App />);

    expect(searchClient.search).toHaveBeenCalledTimes(1);
    expect(searchClient.search).toHaveBeenCalledWith([
      {
        indexName: 'instant_search',
        params: {
          facetFilters: [['brand:Apple']],
          facets: ['brand'],
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          maxValuesPerFacet: 10,
          query: 'iphone',
          tagFilters: '',
        },
      },
      {
        indexName: 'instant_search',
        params: {
          analytics: false,
          attributesToHighlight: [],
          attributesToRetrieve: [],
          attributesToSnippet: [],
          clickAnalytics: false,
          facets: 'brand',
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          hitsPerPage: 1,
          maxValuesPerFacet: 10,
          query: 'iphone',
          page: 0,
          tagFilters: '',
        },
      },
      {
        indexName: 'instant_search_price_asc',
        params: {
          facetFilters: [['brand:Apple']],
          facets: ['brand'],
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          maxValuesPerFacet: 10,
          query: 'iphone',
          tagFilters: '',
        },
      },
      {
        indexName: 'instant_search_price_asc',
        params: {
          analytics: false,
          attributesToHighlight: [],
          attributesToRetrieve: [],
          attributesToSnippet: [],
          clickAnalytics: false,
          facets: 'brand',
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          hitsPerPage: 1,
          maxValuesPerFacet: 10,
          page: 0,
          query: 'iphone',
          tagFilters: '',
        },
      },
      {
        indexName: 'instant_search_rating_desc',
        params: {
          facetFilters: [['brand:Apple']],
          facets: ['brand'],
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          maxValuesPerFacet: 10,
          query: 'iphone',
          tagFilters: '',
        },
      },
      {
        indexName: 'instant_search_rating_desc',
        params: {
          analytics: false,
          attributesToHighlight: [],
          attributesToRetrieve: [],
          attributesToSnippet: [],
          clickAnalytics: false,
          facets: 'brand',
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          hitsPerPage: 1,
          maxValuesPerFacet: 10,
          page: 0,
          query: 'iphone',
          tagFilters: '',
        },
      },
      {
        indexName: 'instant_search_price_desc',
        params: {
          facetFilters: [['brand:Apple']],
          facets: ['brand'],
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          maxValuesPerFacet: 10,
          query: 'iphone',
          tagFilters: '',
        },
      },
      {
        indexName: 'instant_search_price_desc',
        params: {
          analytics: false,
          attributesToHighlight: [],
          attributesToRetrieve: [],
          attributesToSnippet: [],
          clickAnalytics: false,
          facets: 'brand',
          highlightPostTag: '__/ais-highlight__',
          highlightPreTag: '__ais-highlight__',
          hitsPerPage: 1,
          maxValuesPerFacet: 10,
          page: 0,
          query: 'iphone',
          tagFilters: '',
        },
      },
    ]);
  });

  test('returns initialResults', async () => {
    const searchClient = createSearchClient();
    const { App } = createTestEnvironment({ searchClient });

    const serverState = await getServerState(<App />);

    expect(Object.keys(serverState.initialResults)).toHaveLength(4);
    expect(serverState.initialResults).toMatchSnapshot();
  });

  test('searches twice (cached) with dynamic widgets', async () => {
    const searchClient = createSearchClient();
    const { App } = createTestEnvironment({ searchClient, initialUiState: {} });

    await getServerState(
      <App>
        <DynamicWidgets fallbackComponent={RefinementList} />
      </App>
    );

    expect(searchClient.search).toHaveBeenCalledTimes(2);
    // both calls are the same, so they're cached
    expect((searchClient.search as jest.Mock).mock.calls[0][0]).toEqual(
      (searchClient.search as jest.Mock).mock.calls[1][0]
    );
  });

  test('searches twice (cached) with dynamic widgets inside index', async () => {
    const searchClient = createSearchClient();
    const { App } = createTestEnvironment({ searchClient, initialUiState: {} });

    await getServerState(
      <App>
        <Index indexName="something">
          <DynamicWidgets fallbackComponent={RefinementList} />
        </Index>
      </App>
    );

    expect(searchClient.search).toHaveBeenCalledTimes(2);
    // both calls are the same, so they're cached
    expect((searchClient.search as jest.Mock).mock.calls[0][0]).toEqual(
      (searchClient.search as jest.Mock).mock.calls[1][0]
    );
  });

  test('searches twice with dynamic widgets and a refinement', async () => {
    const searchClient = createSearchClient();
    const { App } = createTestEnvironment({
      searchClient,
      initialUiState: {
        instant_search: {
          refinementList: {
            categories: ['refined!'],
          },
        },
      },
    });

    await getServerState(
      <App>
        <DynamicWidgets fallbackComponent={RefinementList} />
      </App>
    );

    expect(searchClient.search).toHaveBeenCalledTimes(2);

    // first query doesn't have the fallback widget mounted yet
    expect((searchClient.search as jest.Mock).mock.calls[0][0][0]).toEqual({
      indexName: 'instant_search',
      params: {
        facets: ['*'],
        highlightPostTag: '__/ais-highlight__',
        highlightPreTag: '__ais-highlight__',
        maxValuesPerFacet: 20,
        query: '',
        tagFilters: '',
      },
    });

    // second query does have the fallback widget mounted, and thus also the refinement
    expect((searchClient.search as jest.Mock).mock.calls[1][0][0]).toEqual({
      indexName: 'instant_search',
      params: {
        facetFilters: [['categories:refined!']],
        facets: ['*'],
        highlightPostTag: '__/ais-highlight__',
        highlightPreTag: '__ais-highlight__',
        maxValuesPerFacet: 20,
        query: '',
        tagFilters: '',
      },
    });
  });
});

function SearchBox() {
  useSearchBox();
  return null;
}

function Hits() {
  useHits();
  return null;
}

function RefinementList(props: UseRefinementListProps) {
  useRefinementList(props);
  return null;
}
