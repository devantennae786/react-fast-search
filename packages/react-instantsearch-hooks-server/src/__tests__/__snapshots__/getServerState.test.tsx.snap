// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getServerState returns initialResults 1`] = `
Object {
  "instant_search": Object {
    "results": Array [
      Object {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": Array [],
        "hitsPerPage": 20,
        "index": "instant_search",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "maxValuesPerFacet=10&query=iphone&highlightPreTag=__ais-highlight__&highlightPostTag=__%2Fais-highlight__&facets=%5B%22brand%22%5D&tagFilters=&facetFilters=%5B%5B%22brand%3AApple%22%5D%5D",
        "processingTimeMS": 0,
        "query": "",
        "renderingContent": Object {
          "facetOrdering": Object {
            "facets": Object {
              "order": Array [
                "brand",
                "hierarchicalCategories.lvl0",
                "categories",
              ],
            },
            "values": Object {
              "brand": Object {
                "sortRemainingBy": "count",
              },
              "categories": Object {
                "sortRemainingBy": "count",
              },
              "hierarchicalCategories.lvl0": Object {
                "sortRemainingBy": "count",
              },
            },
          },
        },
        "userData": Array [
          Object {
            "banner": "https://banner.jpg",
            "link": "https://banner.com/link/",
            "title": "Banner title",
          },
        ],
      },
      Object {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": Array [],
        "hitsPerPage": 20,
        "index": "instant_search",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "maxValuesPerFacet=10&query=iphone&highlightPreTag=__ais-highlight__&highlightPostTag=__%2Fais-highlight__&hitsPerPage=1&page=0&attributesToRetrieve=%5B%5D&attributesToHighlight=%5B%5D&attributesToSnippet=%5B%5D&tagFilters=&analytics=false&clickAnalytics=false&facets=brand",
        "processingTimeMS": 0,
        "query": "",
        "renderingContent": Object {
          "facetOrdering": Object {
            "facets": Object {
              "order": Array [
                "brand",
                "hierarchicalCategories.lvl0",
                "categories",
              ],
            },
            "values": Object {
              "brand": Object {
                "sortRemainingBy": "count",
              },
              "categories": Object {
                "sortRemainingBy": "count",
              },
              "hierarchicalCategories.lvl0": Object {
                "sortRemainingBy": "count",
              },
            },
          },
        },
        "userData": Array [
          Object {
            "banner": "https://banner.jpg",
            "link": "https://banner.com/link/",
            "title": "Banner title",
          },
        ],
      },
    ],
    "state": Object {
      "disjunctiveFacets": Array [
        "brand",
      ],
      "disjunctiveFacetsRefinements": Object {
        "brand": Array [
          "Apple",
        ],
      },
      "facets": Array [],
      "facetsExcludes": Object {},
      "facetsRefinements": Object {},
      "hierarchicalFacets": Array [],
      "hierarchicalFacetsRefinements": Object {},
      "highlightPostTag": "__/ais-highlight__",
      "highlightPreTag": "__ais-highlight__",
      "index": "instant_search",
      "maxValuesPerFacet": 10,
      "numericRefinements": Object {},
      "query": "iphone",
      "tagRefinements": Array [],
    },
  },
  "instant_search_price_asc": Object {
    "results": Array [
      Object {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": Array [],
        "hitsPerPage": 20,
        "index": "instant_search_price_asc",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "maxValuesPerFacet=10&query=iphone&highlightPreTag=__ais-highlight__&highlightPostTag=__%2Fais-highlight__&facets=%5B%22brand%22%5D&tagFilters=&facetFilters=%5B%5B%22brand%3AApple%22%5D%5D",
        "processingTimeMS": 0,
        "query": "",
        "renderingContent": Object {
          "facetOrdering": Object {
            "facets": Object {
              "order": Array [
                "brand",
                "hierarchicalCategories.lvl0",
                "categories",
              ],
            },
            "values": Object {
              "brand": Object {
                "sortRemainingBy": "count",
              },
              "categories": Object {
                "sortRemainingBy": "count",
              },
              "hierarchicalCategories.lvl0": Object {
                "sortRemainingBy": "count",
              },
            },
          },
        },
        "userData": Array [
          Object {
            "banner": "https://banner.jpg",
            "link": "https://banner.com/link/",
            "title": "Banner title",
          },
        ],
      },
      Object {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": Array [],
        "hitsPerPage": 20,
        "index": "instant_search_price_asc",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "maxValuesPerFacet=10&query=iphone&highlightPreTag=__ais-highlight__&highlightPostTag=__%2Fais-highlight__&hitsPerPage=1&page=0&attributesToRetrieve=%5B%5D&attributesToHighlight=%5B%5D&attributesToSnippet=%5B%5D&tagFilters=&analytics=false&clickAnalytics=false&facets=brand",
        "processingTimeMS": 0,
        "query": "",
        "renderingContent": Object {
          "facetOrdering": Object {
            "facets": Object {
              "order": Array [
                "brand",
                "hierarchicalCategories.lvl0",
                "categories",
              ],
            },
            "values": Object {
              "brand": Object {
                "sortRemainingBy": "count",
              },
              "categories": Object {
                "sortRemainingBy": "count",
              },
              "hierarchicalCategories.lvl0": Object {
                "sortRemainingBy": "count",
              },
            },
          },
        },
        "userData": Array [
          Object {
            "banner": "https://banner.jpg",
            "link": "https://banner.com/link/",
            "title": "Banner title",
          },
        ],
      },
    ],
    "state": Object {
      "disjunctiveFacets": Array [
        "brand",
      ],
      "disjunctiveFacetsRefinements": Object {
        "brand": Array [
          "Apple",
        ],
      },
      "facets": Array [],
      "facetsExcludes": Object {},
      "facetsRefinements": Object {},
      "hierarchicalFacets": Array [],
      "hierarchicalFacetsRefinements": Object {},
      "highlightPostTag": "__/ais-highlight__",
      "highlightPreTag": "__ais-highlight__",
      "index": "instant_search_price_asc",
      "maxValuesPerFacet": 10,
      "numericRefinements": Object {},
      "query": "iphone",
      "tagRefinements": Array [],
    },
  },
  "instant_search_price_desc": Object {
    "results": Array [
      Object {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": Array [],
        "hitsPerPage": 20,
        "index": "instant_search_price_desc",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "maxValuesPerFacet=10&query=iphone&highlightPreTag=__ais-highlight__&highlightPostTag=__%2Fais-highlight__&facets=%5B%22brand%22%5D&tagFilters=&facetFilters=%5B%5B%22brand%3AApple%22%5D%5D",
        "processingTimeMS": 0,
        "query": "",
        "renderingContent": Object {
          "facetOrdering": Object {
            "facets": Object {
              "order": Array [
                "brand",
                "hierarchicalCategories.lvl0",
                "categories",
              ],
            },
            "values": Object {
              "brand": Object {
                "sortRemainingBy": "count",
              },
              "categories": Object {
                "sortRemainingBy": "count",
              },
              "hierarchicalCategories.lvl0": Object {
                "sortRemainingBy": "count",
              },
            },
          },
        },
        "userData": Array [
          Object {
            "banner": "https://banner.jpg",
            "link": "https://banner.com/link/",
            "title": "Banner title",
          },
        ],
      },
      Object {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": Array [],
        "hitsPerPage": 20,
        "index": "instant_search_price_desc",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "maxValuesPerFacet=10&query=iphone&highlightPreTag=__ais-highlight__&highlightPostTag=__%2Fais-highlight__&hitsPerPage=1&page=0&attributesToRetrieve=%5B%5D&attributesToHighlight=%5B%5D&attributesToSnippet=%5B%5D&tagFilters=&analytics=false&clickAnalytics=false&facets=brand",
        "processingTimeMS": 0,
        "query": "",
        "renderingContent": Object {
          "facetOrdering": Object {
            "facets": Object {
              "order": Array [
                "brand",
                "hierarchicalCategories.lvl0",
                "categories",
              ],
            },
            "values": Object {
              "brand": Object {
                "sortRemainingBy": "count",
              },
              "categories": Object {
                "sortRemainingBy": "count",
              },
              "hierarchicalCategories.lvl0": Object {
                "sortRemainingBy": "count",
              },
            },
          },
        },
        "userData": Array [
          Object {
            "banner": "https://banner.jpg",
            "link": "https://banner.com/link/",
            "title": "Banner title",
          },
        ],
      },
    ],
    "state": Object {
      "disjunctiveFacets": Array [
        "brand",
      ],
      "disjunctiveFacetsRefinements": Object {
        "brand": Array [
          "Apple",
        ],
      },
      "facets": Array [],
      "facetsExcludes": Object {},
      "facetsRefinements": Object {},
      "hierarchicalFacets": Array [],
      "hierarchicalFacetsRefinements": Object {},
      "highlightPostTag": "__/ais-highlight__",
      "highlightPreTag": "__ais-highlight__",
      "index": "instant_search_price_desc",
      "maxValuesPerFacet": 10,
      "numericRefinements": Object {},
      "query": "iphone",
      "tagRefinements": Array [],
    },
  },
  "instant_search_rating_desc": Object {
    "results": Array [
      Object {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": Array [],
        "hitsPerPage": 20,
        "index": "instant_search_rating_desc",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "maxValuesPerFacet=10&query=iphone&highlightPreTag=__ais-highlight__&highlightPostTag=__%2Fais-highlight__&facets=%5B%22brand%22%5D&tagFilters=&facetFilters=%5B%5B%22brand%3AApple%22%5D%5D",
        "processingTimeMS": 0,
        "query": "",
        "renderingContent": Object {
          "facetOrdering": Object {
            "facets": Object {
              "order": Array [
                "brand",
                "hierarchicalCategories.lvl0",
                "categories",
              ],
            },
            "values": Object {
              "brand": Object {
                "sortRemainingBy": "count",
              },
              "categories": Object {
                "sortRemainingBy": "count",
              },
              "hierarchicalCategories.lvl0": Object {
                "sortRemainingBy": "count",
              },
            },
          },
        },
        "userData": Array [
          Object {
            "banner": "https://banner.jpg",
            "link": "https://banner.com/link/",
            "title": "Banner title",
          },
        ],
      },
      Object {
        "exhaustiveFacetsCount": true,
        "exhaustiveNbHits": true,
        "hits": Array [],
        "hitsPerPage": 20,
        "index": "instant_search_rating_desc",
        "nbHits": 0,
        "nbPages": 0,
        "page": 0,
        "params": "maxValuesPerFacet=10&query=iphone&highlightPreTag=__ais-highlight__&highlightPostTag=__%2Fais-highlight__&hitsPerPage=1&page=0&attributesToRetrieve=%5B%5D&attributesToHighlight=%5B%5D&attributesToSnippet=%5B%5D&tagFilters=&analytics=false&clickAnalytics=false&facets=brand",
        "processingTimeMS": 0,
        "query": "",
        "renderingContent": Object {
          "facetOrdering": Object {
            "facets": Object {
              "order": Array [
                "brand",
                "hierarchicalCategories.lvl0",
                "categories",
              ],
            },
            "values": Object {
              "brand": Object {
                "sortRemainingBy": "count",
              },
              "categories": Object {
                "sortRemainingBy": "count",
              },
              "hierarchicalCategories.lvl0": Object {
                "sortRemainingBy": "count",
              },
            },
          },
        },
        "userData": Array [
          Object {
            "banner": "https://banner.jpg",
            "link": "https://banner.com/link/",
            "title": "Banner title",
          },
        ],
      },
    ],
    "state": Object {
      "disjunctiveFacets": Array [
        "brand",
      ],
      "disjunctiveFacetsRefinements": Object {
        "brand": Array [
          "Apple",
        ],
      },
      "facets": Array [],
      "facetsExcludes": Object {},
      "facetsRefinements": Object {},
      "hierarchicalFacets": Array [],
      "hierarchicalFacetsRefinements": Object {},
      "highlightPostTag": "__/ais-highlight__",
      "highlightPreTag": "__ais-highlight__",
      "index": "instant_search_rating_desc",
      "maxValuesPerFacet": 10,
      "numericRefinements": Object {},
      "query": "iphone",
      "tagRefinements": Array [],
    },
  },
}
`;
